src/runtime_src/tools/xclbinutil/XclBinUtilities.cxx

```diff
@@ -595,6 +596,241 @@ XclBinUtilities::write_htonl(std::ostream & _buf, uint32_t _word32)
+            });
+
+  // Determine and recode the grouped memory range 
+  for (unsigned int index = 0; index < workingConnections.size(); ++index) {
+    const unsigned int startIndex = index;
+    unsigned int endIndex = index;
+    const uint64_t groupBaseAddress = workingConnections[startIndex].baseAddress;
+    uint64_t groupSize = workingConnections[startIndex].size;
+
+    // Peek at the next entry
+    for ( ; endIndex + 1 < workingConnections.size(); ++endIndex) {
+      const unsigned int peekIndex = endIndex + 1;
+      const uint64_t nextBaseAddress = groupBaseAddress + groupSize;
+      if ((nextBaseAddress != workingConnections[peekIndex].baseAddress) ||
+          (workingConnections[startIndex].memType.compare(workingConnections[peekIndex].memType)) ||
+          (workingConnections[startIndex].ipLayoutIndex != workingConnections[peekIndex].ipLayoutIndex) ||
+          (workingConnections[startIndex].argIndex != workingConnections[peekIndex].argIndex)) 
+        break;
+      groupSize += workingConnections[endIndex +1].size;
+    }
+    
+    // Update to our next working index
+    index = endIndex;
+        
+    // If range is 1 then no grouping is needed
+    if (startIndex == endIndex) {
+      addConnection(groupConnectivity, workingConnections[startIndex].argIndex, workingConnections[startIndex].ipLayoutIndex, workingConnections[startIndex].memIndex);
+      continue;
+    }
+
+    // Create a group entry based on the first memory entry
+    boost::property_tree::ptree ptGroupMemory = groupTopology[workingConnections[startIndex].memIndex];
+
+    // Update size
+    const boost::optional<std::string> sSizeBytes = ptGroupMemory.get_optional<std::string>("m_size");
+    if (sSizeBytes.is_initialized()) 
+      ptGroupMemory.put("m_size", XUtil::format("0x%lx", groupSize).c_str());
+    else 
+      ptGroupMemory.put("m_sizeKB", XUtil::format("0x%lx", groupSize / 1024).c_str());
+
+    // Add a tag value to indicate that this entry was the result of grouping memories
+    const std::string newTag = ptGroupMemory.get<std::string>("m_tag") + "-G";
+    if (newTag.size() < sizeof(mem_data::m_tag) )
+      ptGroupMemory.put("m_tag", newTag.c_str());
+
+    // Add memory topology entry
+    const unsigned int groupMemIndex = (unsigned int) groupTopology.size();
+    groupTopology.push_back(ptGroupMemory);
+
+    // Create the connection entry
+    addConnection(groupConnectivity, workingConnections[startIndex].argIndex, workingConnections[startIndex].ipLayoutIndex, groupMemIndex);
+  }
+}
+
+static void
+transformMemoryBankGroupingCollections(const std::vector<boost::property_tree::ptree> & connectivity,
+                                       std::vector<boost::property_tree::ptree> & groupTopology, 
+                                       std::vector<boost::property_tree::ptree> & groupConnectivity)
+{
+  // Memory types that can be grouped
+  static const std::vector<std::string> validGroupTypes = { "MEM_HBM" };
```

Comment:
do we really care about the type of memory or it can be grouped as long as the endpoint address is contiguous and connects to one CU arg?

Reply:
Working with Max and Soren offline, for this release, we will only support contiguous groupings of memory for a homogeneous memory configuration.